include/gridtools/communication/GCL.hpp

```diff
@@ -9,63 +9,35 @@
  */
 #pragma once
 
-#include <iostream>
-#ifdef GCL_MPI
 #include <mpi.h>
-#endif
-
-#ifdef GCL_GPU
-#include "../common/cuda_util.hpp"
-#endif
-#include "../common/host_device.hpp"
-
-#ifdef GCL_TRACE
-#include "high_level/stats_collector.hpp"
-#endif
-
-#include "low_level/gcl_arch.hpp"
-
-#ifdef GCL_GPU
-#ifdef GCL_MULTI_STREAMS
-#ifdef GCL_USE_3
-extern cudaStream_t ZL_stream;
-extern cudaStream_t &ZU_stream;
-extern cudaStream_t YL_stream;
-extern cudaStream_t &YU_stream;
-extern cudaStream_t XL_stream;
-extern cudaStream_t &XU_stream;
-#else
-extern cudaStream_t ZL_stream;
-extern cudaStream_t ZU_stream;
-extern cudaStream_t YL_stream;
-extern cudaStream_t YU_stream;
-extern cudaStream_t XL_stream;
-extern cudaStream_t XU_stream;
-#endif
-#else
-#define ZL_stream 0
-#define ZU_stream 0
-#define YL_stream 0
-#define YU_stream 0
-#define XL_stream 0
-#define XU_stream 0
-#endif
-#endif
 
 namespace gridtools {
-
-#ifdef GCL_MPI
-    extern MPI_Comm GCL_WORLD;
-#else
-    extern int GCL_WORLD;
-#endif
-    extern int PID;
-    extern int PROCS;
-
-    void GCL_Init(int argc, char **argv);
-
-    void GCL_Init();
-
-    void GCL_Finalize();
-
+    namespace gcl_impl_ {
+        inline int &pid_holder() {
+            static int res;
+            return res;
+        }
+
+        inline int &procs_holder() {
+            static int res;
+            return res;
+        }
+    } // namespace gcl_impl_
+
+    inline auto GCL_world() { return MPI_COMM_WORLD; }
+
+    inline int GCL_pid() { return gcl_impl_::pid_holder(); }
+
+    inline int GCL_procs() { return gcl_impl_::procs_holder(); }
+
+    inline void GCL_Init(int argc, char **argv) {
+        int ready;
```

Comment:
Are we making this simplification because the other optional macros before were not tested?

Reply:
Yes, exactly.

Reply:
Wouldn't it be better to keep GCL as it was instead?

Reply:
There is a severe issue with the GCL in its current state:

 Both interface and implementation is controlled by macros. Macros affect both the binary and the headers. Being compiled, the binary is incompatible with the headers if the different set of macros is used. Even though the same compiler and the same set of external dependencies is used. De facto we have a cartesian product of libraries that all go under the same name. To manage this complexity correctly Gridtools CMake should expose all GCL macros as flags. As 
we claim Gridtools as a production quality project, all legit combinations of those flags should be tested in the nightly build. Note that we can not do those tests within a single CMake configuration because we currently have only one GCL binary target. We are forced to move this complexity to the jenkins level as well.
My overall conclusion is the following: if we don't want to fix the root of the problem (macros that affect both interface and implementation) we will end up with increasing the maintenance burden of the problem.